src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala

```diff
@@ -244,66 +243,56 @@ private[impl] class LaunchQueueActor(
         case None => sender() ! None
       }
 
-    case add @ Add(runSpec, _) =>
+    case add: Add =>
       // we cannot process more Add requests for one runSpec in parallel because it leads to race condition
       // the queue handling is helping us ensure we do only one per run-spec at a time
-      // requests for multiple runSpecs are still processed in parallel
 
-      val oldQueue: Queue[QueuedAdd] = updatesByRunSpecId(runSpec.id)
-      val newQueue = oldQueue :+ QueuedAdd(sender(), add)
-      updatesByRunSpecId += runSpec.id -> newQueue
-
-      if (oldQueue.isEmpty) {
+      if (queuedUpdates.isEmpty) {
         // start processing the just received operation
-        processNextAddIfExists(runSpec)
+        processNextAdd(QueuedAdd(sender(), add))
+      } else {
+        queuedUpdates += QueuedAdd(sender(), add)
       }
 
     case AddFinished(queuedAdd) =>
-      val add = queuedAdd.add
-      val (dequeued, newQueue) = updatesByRunSpecId(add.spec.id).dequeue
-      require(dequeued == queuedAdd)
-      if (newQueue.isEmpty)
-        updatesByRunSpecId -= add.spec.id
-      else
-        updatesByRunSpecId += add.spec.id -> newQueue
-
-      sender() ! Done
+      queuedAdd.sender ! Done
 
-      processNextAddIfExists(add.spec)
+      if (queuedUpdates.nonEmpty) {
+        processNextAdd(queuedUpdates.dequeue())
+      }
 
     case msg @ RateLimiterActor.DelayUpdate(app, _) =>
       launchers.get(app.id).foreach(_.forward(msg))
   }
 
   @SuppressWarnings(Array("all")) /* async/await */
-  private def processNextAddIfExists(runSpec: RunSpec): Unit = {
+  private def processNextAdd(queuedItem: QueuedAdd): Unit = {
```

Comment:
@alenkacz, could you explain the race condition to @zen-dog?

Reply:
@zen-dog if I am not mistaken, it was this one https://jira.mesosphere.com/browse/MARATHON-8320